spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/MessageSerializationUtils.java

```diff
@@ -0,0 +1,171 @@
+	 * @param contentTypeResolver used for resolving the mime type.
+	 * @param payloadTypeCache Map used for caching the payload type
+	 * @param codec used for deserialization
+	 * @return Deserialized Message.
+	 */
+	public static MessageValues deserializePayload(MessageValues messageValues, ContentTypeResolver contentTypeResolver,
+												Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		Object originalPayload = messageValues.getPayload();
+		MimeType contentType = contentTypeResolver.resolve(new MessageHeaders(messageValues.getHeaders()));
+		Object payload = deserializePayload(originalPayload, contentType, payloadTypeCache, codec);
+		if (payload != null) {
+			messageValues.setPayload(payload);
+			Object originalContentType = messageValues.get(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
+			// Reset content-type only if the original content type is not null (when
+			// receiving messages from
+			// non-SCSt applications).
+			if (originalContentType != null) {
+				messageValues.put(MessageHeaders.CONTENT_TYPE, originalContentType);
+				messageValues.remove(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE);
+			}
+		}
+		return messageValues;
+	}
+
+	private static Object deserializePayload(Object payload, MimeType contentType,
+											Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		if (payload instanceof byte[]) {
+			if (contentType == null || MimeTypeUtils.APPLICATION_OCTET_STREAM.equals(contentType)) {
+				return payload;
+			}
+			else {
+				return deserializePayload((byte[]) payload, contentType, payloadTypeCache, codec);
+			}
+		}
+		return payload;
+	}
+
+	private static Object deserializePayload(byte[] bytes, MimeType contentType,
+									Map<String, Class<?>> payloadTypeCache, Codec codec) {
+		if ("text".equalsIgnoreCase(contentType.getType()) || MimeTypeUtils.APPLICATION_JSON.equals(contentType)) {
+			try {
+				return new String(bytes, "UTF-8");
+			}
+			catch (UnsupportedEncodingException e) {
+				String errorMessage = "unable to deserialize [java.lang.String]. Encoding not supported. "
+						+ e.getMessage();
+				MessageSerializationUtils.LOGGER.error(errorMessage);
+				throw new SerializationFailedException(errorMessage, e);
+			}
+		}
+		else {
+			String className = JavaClassMimeTypeUtils.classNameFromMimeType(contentType);
+			try {
+				// Cache types to avoid unnecessary ClassUtils.forName calls.
+				Class<?> targetType = payloadTypeCache.get(className);
+				if (targetType == null) {
+					targetType = ClassUtils.forName(className, null);
+					payloadTypeCache.put(className, targetType);
+				}
+				return codec.decode(bytes, targetType);
+			} // catch all exceptions that could occur during de-serialization
+			catch (Exception e) {
+				String errorMessage = "Unable to deserialize [" + className + "] using the contentType [" + contentType
+						+ "] " + e.getMessage();
+				MessageSerializationUtils.LOGGER.error(errorMessage);
```

Comment:
I wouldn't log error if I throw it.
That is already this method user to decide what to do with the exception, but not utility